From 6668e1fe1744ce69c08d77b2bbb6ec0d740c6d85 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?R=C3=A9mi=20Bernon?= <rbernon@codeweavers.com>
Date: Thu, 16 Jan 2020 12:30:15 +0100
Subject: [PATCH 10/22] winex11.drv: Introduce new (get|set)_hwnd_for_window
 helpers.

Wrapping the Xlib winContext context uses.

CW-Bug-Id: #19553
CW-Bug-Id: #19644
Wine-Bug: https://bugs.winehq.org/show_bug.cgi?id=2155
---
 dlls/winex11.drv/event.c  | 15 +++-----------
 dlls/winex11.drv/window.c | 41 +++++++++++++++++++++++++++++++--------
 dlls/winex11.drv/x11drv.h |  3 +++
 3 files changed, 39 insertions(+), 20 deletions(-)

diff --git a/dlls/winex11.drv/event.c b/dlls/winex11.drv/event.c
index 11111111111..11111111111 100644
--- a/dlls/winex11.drv/event.c
+++ b/dlls/winex11.drv/event.c
@@ -385,9 +385,7 @@ static inline BOOL call_event_handler( Display *display, XEvent *event )
 #ifdef GenericEvent
     if (event->type == GenericEvent) hwnd = 0; else
 #endif
-    if (XFindContext( display, event->xany.window, winContext, (char **)&hwnd ) != 0)
-        hwnd = 0;  /* not for a registered window */
-    if (!hwnd && event->xany.window == root_window) hwnd = NtUserGetDesktopWindow();
+    hwnd = get_hwnd_for_window( display, event->xany.window );
 
     TRACE( "%lu %s for hwnd/window %p/%lx\n",
            event->xany.serial, dbgstr_event( event->type ), hwnd, event->xany.window );
@@ -839,7 +837,6 @@ static BOOL X11DRV_FocusIn( HWND hwnd, XEvent *xev )
  */
 static void focus_out( Display *display , HWND hwnd )
  {
-    HWND hwnd_tmp;
     Window focus_win;
     int revert;
     XIC xic;
@@ -861,13 +858,7 @@ static void focus_out( Display *display , HWND hwnd )
        getting the focus is a Wine window */
 
     XGetInputFocus( display, &focus_win, &revert );
-    if (focus_win)
-    {
-        if (XFindContext( display, focus_win, winContext, (char **)&hwnd_tmp ) != 0)
-            focus_win = 0;
-    }
-
-    if (!focus_win)
+    if (!get_hwnd_for_window( display, focus_win ))
     {
         /* Abey : 6-Oct-99. Check again if the focus out window is the
            Foreground window, because in most cases the messages sent
@@ -1733,7 +1724,7 @@ static void handle_dnd_protocol( HWND hwnd, XClientMessageEvent *event )
     /* query window (drag&drop event contains only drag window) */
     XQueryPointer( event->display, root_window, &root, &child,
                    &root_x, &root_y, &child_x, &child_y, &u);
-    if (XFindContext( event->display, child, winContext, (char **)&hwnd ) != 0) hwnd = 0;
+    hwnd = get_hwnd_for_window( event->display, child );
     if (!hwnd) return;
     if (event->data.l[0] == DndFile || event->data.l[0] == DndFiles)
         EVENT_DropFromOffiX(hwnd, event);
diff --git a/dlls/winex11.drv/window.c b/dlls/winex11.drv/window.c
index 11111111111..11111111111 100644
--- a/dlls/winex11.drv/window.c
+++ b/dlls/winex11.drv/window.c
@@ -1548,7 +1548,7 @@ Window create_client_window( HWND hwnd, const XVisualInfo *visual )
 
     if (data->client_window)
     {
-        XDeleteContext( data->display, data->client_window, winContext );
+        set_hwnd_for_window( data->display, data->client_window, 0 );
         XReparentWindow( gdi_display, data->client_window, dummy_parent, 0, 0 );
         TRACE( "%p reparent xwin %lx/%lx\n", data->hwnd, data->whole_window, data->client_window );
     }
@@ -1576,7 +1576,7 @@ Window create_client_window( HWND hwnd, const XVisualInfo *visual )
                                                CWBackingStore | CWColormap | CWBorderPixel, &attr );
     if (data->client_window)
     {
-        XSaveContext( data->display, data->client_window, winContext, (char *)data->hwnd );
+        set_hwnd_for_window( data->display, data->client_window, data->hwnd );
         XMapWindow( gdi_display, data->client_window );
         XSync( gdi_display, False );
         if (data->whole_window) XSelectInput( data->display, data->client_window, ExposureMask );
@@ -1636,7 +1636,7 @@ static void create_whole_window( struct x11drv_win_data *data )
     set_initial_wm_hints( data->display, data->whole_window );
     set_wm_hints( data );
 
-    XSaveContext( data->display, data->whole_window, winContext, (char *)data->hwnd );
+    set_hwnd_for_window( data->display, data->whole_window, data->hwnd );
     NtUserSetProp( data->hwnd, whole_window_prop, (HANDLE)data->whole_window );
 
     /* set the window text */
@@ -1668,7 +1668,7 @@ static void destroy_whole_window( struct x11drv_win_data *data, BOOL already_des
 {
     TRACE( "win %p xwin %lx/%lx\n", data->hwnd, data->whole_window, data->client_window );
 
-    if (data->client_window) XDeleteContext( data->display, data->client_window, winContext );
+    if (data->client_window) set_hwnd_for_window( data->display, data->client_window, 0 );
 
     if (!data->whole_window)
     {
@@ -1678,7 +1678,7 @@ static void destroy_whole_window( struct x11drv_win_data *data, BOOL already_des
             if (xwin)
             {
                 if (!already_destroyed) XSelectInput( data->display, xwin, 0 );
-                XDeleteContext( data->display, xwin, winContext );
+                set_hwnd_for_window( data->display, xwin, 0 );
                 NtUserRemoveProp( data->hwnd, foreign_window_prop );
             }
             return;
@@ -1692,7 +1692,7 @@ static void destroy_whole_window( struct x11drv_win_data *data, BOOL already_des
             XReparentWindow( data->display, data->client_window, get_dummy_parent(), 0, 0 );
             XSync( data->display, False );
         }
-        XDeleteContext( data->display, data->whole_window, winContext );
+        set_hwnd_for_window( data->display, data->whole_window, 0 );
         if (!already_destroyed) XDestroyWindow( data->display, data->whole_window );
     }
     if (data->whole_colormap) XFreeColormap( data->display, data->whole_colormap );
@@ -2078,7 +2078,7 @@ HWND create_foreign_window( Display *display, Window xwin )
         class_registered = TRUE;
     }
 
-    if (XFindContext( display, xwin, winContext, (char **)&hwnd )) hwnd = 0;
+    hwnd = get_hwnd_for_window( display, xwin );
     if (hwnd) return hwnd;  /* already created */
 
     XSelectInput( display, xwin, StructureNotifyMask );
@@ -2121,7 +2121,7 @@ HWND create_foreign_window( Display *display, Window xwin )
     data->mapped = TRUE;
 
     NtUserSetProp( hwnd, foreign_window_prop, (HANDLE)xwin );
-    XSaveContext( display, xwin, winContext, (char *)data->hwnd );
+    set_hwnd_for_window( display, xwin, data->hwnd );
 
     TRACE( "win %lx parent %p style %08x %s -> hwnd %p\n",
            xwin, parent, style, wine_dbgstr_rect(&data->window_rect), hwnd );
@@ -3210,3 +3210,28 @@ void init_win_context(void)
     win_data_context = XUniqueContext();
     cursor_context = XUniqueContext();
 }
+
+HWND get_hwnd_for_window( Display *display, Window window )
+{
+    HWND hwnd;
+
+    if (!window) return 0;
+    if (window == root_window) return NtUserGetDesktopWindow();
+
+    TRACE( "display %p, window %lx\n", display, window );
+
+    if (!XFindContext( display, window, winContext, (char **)&hwnd )) return hwnd;
+
+    WARN( "Failed to find HWND for window %lx\n", window );
+    return 0;
+}
+
+void set_hwnd_for_window( Display *display, Window window, HWND hwnd )
+{
+    if (!window || window == root_window) return;
+
+    TRACE( "display %p, window %lx, hwnd %p\n", display, window, hwnd );
+
+    if (!hwnd) XDeleteContext( display, window, winContext );
+    else XSaveContext( display, window, winContext, (char *)hwnd );
+}
diff --git a/dlls/winex11.drv/x11drv.h b/dlls/winex11.drv/x11drv.h
index 11111111111..11111111111 100644
--- a/dlls/winex11.drv/x11drv.h
+++ b/dlls/winex11.drv/x11drv.h
@@ -642,6 +642,9 @@ extern Window X11DRV_get_whole_window( HWND hwnd ) DECLSPEC_HIDDEN;
 extern XIC X11DRV_get_ic( HWND hwnd ) DECLSPEC_HIDDEN;
 extern Window get_dummy_parent(void) DECLSPEC_HIDDEN;
 
+extern HWND get_hwnd_for_window( Display *display, Window window ) DECLSPEC_HIDDEN;
+extern void set_hwnd_for_window( Display *display, Window window, HWND hwnd ) DECLSPEC_HIDDEN;
+
 extern void sync_gl_drawable( HWND hwnd, BOOL known_child ) DECLSPEC_HIDDEN;
 extern void set_gl_drawable_parent( HWND hwnd, HWND parent ) DECLSPEC_HIDDEN;
 extern void destroy_gl_drawable( HWND hwnd ) DECLSPEC_HIDDEN;
-- 
2.37.1

